File: lab3/zad2/cpp/dynarr/dynarrtest.cpp
#define BOOST_TEST_MODULE DynArrTests
#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>
#include <random>
#include <algorithm>
#include "../../../zad1/cpp/dynarr.h"

using namespace std;

string generateRandomString(size_t length = 10) {
    static const char alphanum[] =
        "0123456789"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz";
    
    static mt19937 rng(random_device{}());
    uniform_int_distribution<size_t> dist(0, sizeof(alphanum) - 2);
    
    string result;
    result.reserve(length);
    for (size_t i = 0; i < length; ++i) {
        result += alphanum[dist(rng)];
    }
    return result;
}

BOOST_AUTO_TEST_SUITE(DynArrTests)

BOOST_AUTO_TEST_CASE(TestConstructorDestructor) {
    DynArr arr1;
    BOOST_CHECK_EQUAL(arr1.getLength(), 0);
    
    DynArr arr2(20);
    BOOST_CHECK_EQUAL(arr2.getLength(), 0);
    
    DynArr* arr3 = new DynArr();
    BOOST_CHECK_NO_THROW(delete arr3);
}

BOOST_AUTO_TEST_CASE(TestAddToEnd) {
    DynArr arr(2);
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    
    arr.addToEnd(str1);
    BOOST_CHECK_EQUAL(arr.getLength(), 1);
    BOOST_CHECK_EQUAL(arr.getElement(0), str1);
    
    arr.addToEnd(str2);
    BOOST_CHECK_EQUAL(arr.getLength(), 2);
    BOOST_CHECK_EQUAL(arr.getElement(1), str2);
    
    arr.addToEnd(str3);
    BOOST_CHECK_EQUAL(arr.getLength(), 3);
    BOOST_CHECK_EQUAL(arr.getElement(2), str3);
    
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        arr.addToEnd(str);
        BOOST_CHECK_EQUAL(arr.getElement(3 + i), str);
    }
}

BOOST_AUTO_TEST_CASE(TestAddAtIndex) {
    DynArr arr(3);
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    string str4 = generateRandomString();
    string str5 = generateRandomString();
    
    arr.addAtIndex(0, str1);
    BOOST_CHECK_EQUAL(arr.getLength(), 1);
    BOOST_CHECK_EQUAL(arr.getElement(0), str1);
    
    arr.addAtIndex(0, str2);
    BOOST_CHECK_EQUAL(arr.getLength(), 2);
    BOOST_CHECK_EQUAL(arr.getElement(0), str2);
    BOOST_CHECK_EQUAL(arr.getElement(1), str1);
    
    arr.addAtIndex(2, str3);
    BOOST_CHECK_EQUAL(arr.getLength(), 3);
    BOOST_CHECK_EQUAL(arr.getElement(2), str3);
    
    arr.addAtIndex(1, str4);
    BOOST_CHECK_EQUAL(arr.getLength(), 4);
    BOOST_CHECK_EQUAL(arr.getElement(1), str4);
    
    arr.addAtIndex(2, str5);
    BOOST_CHECK_EQUAL(arr.getLength(), 5);
    
    BOOST_CHECK_THROW(arr.addAtIndex(-1, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.addAtIndex(10, generateRandomString()), runtime_error);
}

BOOST_AUTO_TEST_CASE(TestGetElement) {
    DynArr arr;
    vector<string> testStrings;
    
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testStrings.push_back(str);
        arr.addToEnd(str);
    }
    
    for (int i = 0; i < 5; ++i) {
        BOOST_CHECK_EQUAL(arr.getElement(i), testStrings[i]);
    }
    
    BOOST_CHECK_THROW(arr.getElement(-1), runtime_error);
    BOOST_CHECK_THROW(arr.getElement(5), runtime_error);
    BOOST_CHECK_THROW(arr.getElement(100), runtime_error);
}

BOOST_AUTO_TEST_CASE(TestReplaceElement) {
    DynArr arr;
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    string new1 = generateRandomString();
    string new2 = generateRandomString();
    string new3 = generateRandomString();
    
    arr.addToEnd(str1);
    arr.addToEnd(str2);
    arr.addToEnd(str3);
    
    arr.replaceElement(0, new1);
    BOOST_CHECK_EQUAL(arr.getElement(0), new1);
    
    arr.replaceElement(1, new2);
    BOOST_CHECK_EQUAL(arr.getElement(1), new2);
    
    arr.replaceElement(2, new3);
    BOOST_CHECK_EQUAL(arr.getElement(2), new3);
    
    BOOST_CHECK_THROW(arr.replaceElement(-1, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.replaceElement(3, generateRandomString()), runtime_error);
}

BOOST_AUTO_TEST_CASE(TestRemoveAtIndex) {
    DynArr arr;
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    string str4 = generateRandomString();
    string str5 = generateRandomString();
    
    arr.addToEnd(str1);
    arr.addToEnd(str2);
    arr.addToEnd(str3);
    arr.addToEnd(str4);
    arr.addToEnd(str5);
    
    arr.removeAtIndex(2);
    BOOST_CHECK_EQUAL(arr.getLength(), 4);
    BOOST_CHECK_EQUAL(arr.getElement(0), str1);
    BOOST_CHECK_EQUAL(arr.getElement(1), str2);
    BOOST_CHECK_EQUAL(arr.getElement(2), str4);
    BOOST_CHECK_EQUAL(arr.getElement(3), str5);
    
    arr.removeAtIndex(3);
    BOOST_CHECK_EQUAL(arr.getLength(), 3);
    BOOST_CHECK_EQUAL(arr.getElement(0), str1);
    BOOST_CHECK_EQUAL(arr.getElement(1), str2);
    BOOST_CHECK_EQUAL(arr.getElement(2), str4);
    
    arr.removeAtIndex(0);
    BOOST_CHECK_EQUAL(arr.getLength(), 2);
    BOOST_CHECK_EQUAL(arr.getElement(0), str2);
    BOOST_CHECK_EQUAL(arr.getElement(1), str4);
    
    arr.removeAtIndex(0);
    BOOST_CHECK_EQUAL(arr.getLength(), 1);
    BOOST_CHECK_EQUAL(arr.getElement(0), str4);
    
    arr.removeAtIndex(0);
    BOOST_CHECK_EQUAL(arr.getLength(), 0);
    
    BOOST_CHECK_THROW(arr.removeAtIndex(-1), runtime_error);
    BOOST_CHECK_THROW(arr.removeAtIndex(0), runtime_error);
}

BOOST_AUTO_TEST_CASE(TestPrintArr) {
    DynArr arr;
    BOOST_CHECK_NO_THROW(arr.printArr());
    
    arr.addToEnd(generateRandomString());
    arr.addToEnd(generateRandomString());
    BOOST_CHECK_NO_THROW(arr.printArr());
}

BOOST_AUTO_TEST_CASE(TestFileOperations) {
    const string filename = "test_dynarr.txt";
    DynArr arr1, arr2;
    
    vector<string> testData;
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testData.push_back(str);
        arr1.addToEnd(str);
    }
    
    ofstream outFile(filename);
    BOOST_REQUIRE(outFile.is_open());
    arr1.saveToFile(outFile);
    outFile.close();
    
    ifstream inFile(filename);
    BOOST_REQUIRE(inFile.is_open());
    arr2.loadFromFile(inFile);
    inFile.close();
    
    BOOST_CHECK_EQUAL(arr1.getLength(), arr2.getLength());
    for (int i = 0; i < arr1.getLength(); ++i) {
        BOOST_CHECK_EQUAL(arr1.getElement(i), arr2.getElement(i));
    }
    
    remove(filename.c_str());
}

BOOST_AUTO_TEST_CASE(TestBinaryFileOperations) {
    const string filename = "test_dynarr_binary.bin";
    DynArr arr1, arr2;
    
    vector<string> testData;
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testData.push_back(str);
        arr1.addToEnd(str);
    }
    
    ofstream outFile(filename, ios::binary);
    BOOST_REQUIRE(outFile.is_open());
    arr1.saveToBinaryFile(outFile);
    outFile.close();
    
    ifstream inFile(filename, ios::binary);
    BOOST_REQUIRE(inFile.is_open());
    arr2.loadFromBinaryFile(inFile);
    inFile.close();
    
    BOOST_CHECK_EQUAL(arr1.getLength(), arr2.getLength());
    for (int i = 0; i < arr1.getLength(); ++i) {
        BOOST_CHECK_EQUAL(arr1.getElement(i), arr2.getElement(i));
    }
    
    remove(filename.c_str());
}

BOOST_AUTO_TEST_CASE(TestExceptionSafety) {
    DynArr arr;
    
    BOOST_CHECK_THROW(arr.getElement(0), runtime_error);
    BOOST_CHECK_THROW(arr.replaceElement(0, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.removeAtIndex(0), runtime_error);
    
    string testStr = generateRandomString();
    arr.addToEnd(testStr);
    
    BOOST_CHECK_THROW(arr.getElement(-1), runtime_error);
    BOOST_CHECK_THROW(arr.getElement(1), runtime_error);
    BOOST_CHECK_THROW(arr.replaceElement(-1, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.replaceElement(1, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.removeAtIndex(-1), runtime_error);
    BOOST_CHECK_THROW(arr.removeAtIndex(1), runtime_error);
    
    BOOST_CHECK_THROW(arr.addAtIndex(-1, generateRandomString()), runtime_error);
    BOOST_CHECK_THROW(arr.addAtIndex(2, generateRandomString()), runtime_error);
}

BOOST_AUTO_TEST_SUITE_END()

File: lab3/zad2/go/dynarr/dynarr_test.go
package dynarr

import (
	"os"
	"testing"

	"github.com/stretchr/testify/suite"
)

type DynArrTestSuite struct {
	suite.Suite
}

func TestDynArrTestSuite(t *testing.T) {
	suite.Run(t, new(DynArrTestSuite))
}

func generateRandomString(length int) string {
	const charset = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	result := make([]byte, length)
	for i := range result {
		result[i] = charset[i%len(charset)]
	}
	return string(result)
}

func (s *DynArrTestSuite) TestConstructor() {
	arr1 := NewDynArr(0)
	s.Equal(DefaultInitialCapacity, arr1.cap)
	s.Equal(0, arr1.len)

	arr2 := NewDynArr(20)
	s.Equal(20, arr2.cap)
	s.Equal(0, arr2.len)

	s.NotPanics(func() {
		_ = NewDynArr(10)
	})
}

func (s *DynArrTestSuite) TestAddToEnd() {
	arr := NewDynArr(2)
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)

	arr.AddToEnd(str1)
	s.Equal(1, arr.len)
	element, _ := arr.GetElement(0)
	s.Equal(str1, element)

	arr.AddToEnd(str2)
	s.Equal(2, arr.len)
	element, _ = arr.GetElement(1)
	s.Equal(str2, element)

	arr.AddToEnd(str3)
	s.Equal(3, arr.len)
	s.Equal(4, arr.cap)
	element, _ = arr.GetElement(2)
	s.Equal(str3, element)

	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		arr.AddToEnd(str)
		element, _ := arr.GetElement(3 + i)
		s.Equal(str, element)
	}
}

func (s *DynArrTestSuite) TestAddAtIndex() {
	arr := NewDynArr(3)
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)
	str4 := generateRandomString(10)
	str5 := generateRandomString(10)

	arr.AddAtIndex(0, str1)
	s.Equal(1, arr.len)
	element, _ := arr.GetElement(0)
	s.Equal(str1, element)

	arr.AddAtIndex(0, str2)
	s.Equal(2, arr.len)
	element, _ = arr.GetElement(0)
	s.Equal(str2, element)
	element, _ = arr.GetElement(1)
	s.Equal(str1, element)

	arr.AddAtIndex(2, str3)
	s.Equal(3, arr.len)
	element, _ = arr.GetElement(2)
	s.Equal(str3, element)

	arr.AddAtIndex(1, str4)
	s.Equal(4, arr.len)
	element, _ = arr.GetElement(1)
	s.Equal(str4, element)

	arr.AddAtIndex(2, str5)
	s.Equal(5, arr.len)

	err := arr.AddAtIndex(-1, generateRandomString(10))
	s.Error(err)

	err = arr.AddAtIndex(10, generateRandomString(10))
	s.Error(err)
}

func (s *DynArrTestSuite) TestGetElement() {
	arr := NewDynArr(2)
	testStrings := []string{}

	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testStrings = append(testStrings, str)
		arr.AddToEnd(str)
	}

	for i := 0; i < 5; i++ {
		element, err := arr.GetElement(i)
		s.NoError(err)
		s.Equal(testStrings[i], element)
	}

	_, err := arr.GetElement(-1)
	s.Error(err)

	_, err = arr.GetElement(5)
	s.Error(err)

	_, err = arr.GetElement(100)
	s.Error(err)
}

func (s *DynArrTestSuite) TestReplaceElement() {
	arr := NewDynArr(3)
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)
	new1 := generateRandomString(10)
	new2 := generateRandomString(10)
	new3 := generateRandomString(10)

	arr.AddToEnd(str1)
	arr.AddToEnd(str2)
	arr.AddToEnd(str3)

	arr.ReplaceElement(0, new1)
	element, _ := arr.GetElement(0)
	s.Equal(new1, element)

	arr.ReplaceElement(1, new2)
	element, _ = arr.GetElement(1)
	s.Equal(new2, element)

	arr.ReplaceElement(2, new3)
	element, _ = arr.GetElement(2)
	s.Equal(new3, element)

	err := arr.ReplaceElement(-1, generateRandomString(10))
	s.Error(err)

	err = arr.ReplaceElement(3, generateRandomString(10))
	s.Error(err)
}

func (s *DynArrTestSuite) TestRemoveAtIndex() {
	arr := NewDynArr(5)
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)
	str4 := generateRandomString(10)
	str5 := generateRandomString(10)

	arr.AddToEnd(str1)
	arr.AddToEnd(str2)
	arr.AddToEnd(str3)
	arr.AddToEnd(str4)
	arr.AddToEnd(str5)

	arr.RemoveAtIndex(2)
	s.Equal(4, arr.len)
	element, _ := arr.GetElement(0)
	s.Equal(str1, element)
	element, _ = arr.GetElement(1)
	s.Equal(str2, element)
	element, _ = arr.GetElement(2)
	s.Equal(str4, element)
	element, _ = arr.GetElement(3)
	s.Equal(str5, element)

	arr.RemoveAtIndex(3)
	s.Equal(3, arr.len)
	element, _ = arr.GetElement(0)
	s.Equal(str1, element)
	element, _ = arr.GetElement(1)
	s.Equal(str2, element)
	element, _ = arr.GetElement(2)
	s.Equal(str4, element)

	arr.RemoveAtIndex(0)
	s.Equal(2, arr.len)
	element, _ = arr.GetElement(0)
	s.Equal(str2, element)
	element, _ = arr.GetElement(1)
	s.Equal(str4, element)

	arr.RemoveAtIndex(0)
	s.Equal(1, arr.len)
	element, _ = arr.GetElement(0)
	s.Equal(str4, element)

	arr.RemoveAtIndex(0)
	s.Equal(0, arr.len)

	err := arr.RemoveAtIndex(-1)
	s.Error(err)

	err = arr.RemoveAtIndex(0)
	s.Error(err)
}

func (s *DynArrTestSuite) TestPrintArr() {
	arr := NewDynArr(2)
	s.NotPanics(arr.PrintArr)

	arr.AddToEnd(generateRandomString(10))
	arr.AddToEnd(generateRandomString(10))
	s.NotPanics(arr.PrintArr)
}

func (s *DynArrTestSuite) TestFileOperations() {
	filename := "test_dynarr.txt"
	arr1 := NewDynArr(2)
	arr2 := NewDynArr(2)

	testData := []string{}
	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testData = append(testData, str)
		arr1.AddToEnd(str)
	}

	file, err := os.Create(filename)
	s.NoError(err)

	err = arr1.SaveToFile(file)
	s.NoError(err)
	file.Close()

	file, err = os.Open(filename)
	s.NoError(err)

	err = arr2.LoadFromFile(file)
	s.NoError(err)
	file.Close()

	s.Equal(arr1.GetLength(), arr2.GetLength())
	for i := 0; i < arr1.GetLength(); i++ {
		element1, _ := arr1.GetElement(i)
		element2, _ := arr2.GetElement(i)
		s.Equal(element1, element2)
	}

	os.Remove(filename)
}

func (s *DynArrTestSuite) TestBinaryFileOperations() {
	filename := "test_dynarr_binary.bin"
	arr1 := NewDynArr(2)
	arr2 := NewDynArr(2)

	testData := []string{}
	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testData = append(testData, str)
		arr1.AddToEnd(str)
	}

	file, err := os.Create(filename)
	s.NoError(err)

	err = arr1.SaveToBinaryFile(file)
	s.NoError(err)
	file.Close()

	file, err = os.Open(filename)
	s.NoError(err)

	err = arr2.LoadFromBinaryFile(file)
	s.NoError(err)
	file.Close()

	s.Equal(arr1.GetLength(), arr2.GetLength())
	for i := 0; i < arr1.GetLength(); i++ {
		element1, _ := arr1.GetElement(i)
		element2, _ := arr2.GetElement(i)
		s.Equal(element1, element2)
	}

	os.Remove(filename)
}

func (s *DynArrTestSuite) TestExceptionSafety() {
	arr := NewDynArr(2)

	_, err := arr.GetElement(0)
	s.Error(err)

	err = arr.ReplaceElement(0, generateRandomString(10))
	s.Error(err)

	err = arr.RemoveAtIndex(0)
	s.Error(err)

	testStr := generateRandomString(10)
	arr.AddToEnd(testStr)

	_, err = arr.GetElement(-1)
	s.Error(err)

	_, err = arr.GetElement(1)
	s.Error(err)

	err = arr.ReplaceElement(-1, generateRandomString(10))
	s.Error(err)

	err = arr.ReplaceElement(1, generateRandomString(10))
	s.Error(err)

	err = arr.RemoveAtIndex(-1)
	s.Error(err)

	err = arr.RemoveAtIndex(1)
	s.Error(err)

	err = arr.AddAtIndex(-1, generateRandomString(10))
	s.Error(err)

	err = arr.AddAtIndex(2, generateRandomString(10))
	s.Error(err)
}

func (s *DynArrTestSuite) TestNilFile() {
	arr := NewDynArr(2)
	arr.AddToEnd("test")

	err := arr.SaveToFile(nil)
	s.Error(err)

	err = arr.LoadFromFile(nil)
	s.Error(err)

	err = arr.SaveToBinaryFile(nil)
	s.Error(err)

	err = arr.LoadFromBinaryFile(nil)
	s.Error(err)
}
